packages/Popover/src/Popover.tsx

```diff
@@ -0,0 +1,626 @@
+    const contentEl = this.contentRef && this.contentRef.current;
+
+    if (contentEl && !contentElPos) {
+      this.setState({
+        contentElPos: this.getRefPosition(contentEl),
+      });
+    }
+  }
+
+  // Gets top offset, left offset, width and height dimensions for a node
+  getRefPosition(element: HTMLElement | null) {
+    if (!element) {
+      return defaultRefPosition;
+    }
+
+    const { top, bottom, left, right } = element.getBoundingClientRect();
+    const { offsetHeight: height, offsetWidth: width } = element;
+
+    return { top, bottom, left, right, height, width };
+  }
+
+  // Returns the style object that is used to position and transition the popover component
+  calculatePosition() {
+    const { withoutPortal } = this.props;
+    const { referenceElement, contentElPos } = this.state;
+
+    // Forced second render to make sure that
+    // we have access to refs
+    if (!this.state.hasMounted) {
+      return;
+    }
+
+    if (!referenceElement) {
+      this.setReferenceElement();
+      return;
+    }
+
+    if (!contentElPos) {
+      this.setContentElPosition();
+      return;
+    }
+
+    const alignment = this.getAlignment();
+    const justification = this.getJustification(alignment);
+
+    const transformOrigin = this.getTransformOrigin({
+      alignment,
+      justification,
+    });
+
+    const transform = this.getTransform(alignment);
+
+    if (withoutPortal) {
+      return {
+        ...this.calcPositionWithoutPortal({ alignment, justification }),
+        transformOrigin,
+        transform,
+      };
+    }
+
+    return {
+      top: this.calcTop({ alignment, justification }),
+      left: this.calcLeft({ alignment, justification }),
+      transformOrigin,
+      transform,
+    };
+  }
+
+  // Determines the alignment to render based on an order of alignment fallbacks
+  // Returns the first alignment that doesn't collide with the window,
+  // defaulting to the align prop if all alignments fail.
+  getAlignment() {
+    const { align } = this.props;
+
+    const alignments: {
+      top: Array<Alignment>;
+      bottom: Array<Alignment>;
+      left: Array<Alignment>;
+      right: Array<Alignment>;
+    } = {
+      top: ['top', 'bottom'],
+      bottom: ['bottom', 'top'],
+      left: ['left', 'right'],
+      right: ['right', 'left'],
+    };
+
+    return (
+      alignments[align].find(candidate => this.checkAlignment(candidate)) ||
+      align
+    );
+  }
+
+  // Determines the justification to render based on an order of justification fallbacks
+  // Returns the first justification that doesn't collide with the window,
+  // defaulting to the justify prop if all justifications fail.
+  getJustification(alignment: Alignment) {
+    const { justify } = this.props;
+
+    let justifications: {
+      start: Array<Justification>;
+      middle: Array<Justification>;
+      end: Array<Justification>;
+    };
+
+    switch (alignment) {
+      case 'left':
```

Comment:
Can then use Enum here when it's defined.

Reply:
do you mean doing Align.left, based on the enum? @hswolff